policy/options.go

```diff
@@ -0,0 +1,460 @@
+package policy
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"golang.org/x/net/idna"
+)
+
+type NamePolicyOption func(e *NamePolicyEngine) error
+
+// TODO: wrap (more) errors; and prove a set of known (exported) errors
+
+func WithSubjectCommonNameVerification() NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		e.verifySubjectCommonName = true
+		return nil
+	}
+}
+
+func WithAllowLiteralWildcardNames() NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		e.allowLiteralWildcardNames = true
+		return nil
+	}
+}
+
+func WithPermittedDNSDomains(domains []string) NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		normalizedDomains := make([]string, len(domains))
+		for i, domain := range domains {
+			normalizedDomain, err := normalizeAndValidateDNSDomainConstraint(domain)
+			if err != nil {
+				return fmt.Errorf("cannot parse permitted domain constraint %q: %w", domain, err)
+			}
+			normalizedDomains[i] = normalizedDomain
+		}
+		e.permittedDNSDomains = normalizedDomains
+		return nil
+	}
+}
+
+func WithExcludedDNSDomains(domains []string) NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		normalizedDomains := make([]string, len(domains))
+		for i, domain := range domains {
+			normalizedDomain, err := normalizeAndValidateDNSDomainConstraint(domain)
+			if err != nil {
+				return fmt.Errorf("cannot parse excluded domain constraint %q: %w", domain, err)
+			}
+			normalizedDomains[i] = normalizedDomain
+		}
+		e.excludedDNSDomains = normalizedDomains
+		return nil
+	}
+}
+
+func WithPermittedDNSDomain(domain string) NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		normalizedDomain, err := normalizeAndValidateDNSDomainConstraint(domain)
+		if err != nil {
+			return fmt.Errorf("cannot parse permitted domain constraint %q: %w", domain, err)
+		}
+		e.permittedDNSDomains = []string{normalizedDomain}
+		return nil
+	}
+}
+
+func WithExcludedDNSDomain(domain string) NamePolicyOption {
```

Comment:
We can get rid of these "duplicated" methods if we change the ones that get a slice with a variadic function, e.g., `func WithExcludedDNSDomains(domains ...string) NamePolicyOption`.

Reply:
Great suggestion. Fixed in a2cfbe3d546404932ca15bd8b7db981932a9a179.